### ISSUE TITLE
expanding MatrixExpr

### ISSUE THREAD

**Issue by goretkin** [diofant/diofant#468]
```
In [1]: import diofant

In [2]: diofant.__version__
Out[2]: '0.9.0.dev0'

In [3]: n = diofant.Symbol("n")

In [4]: A, B, C = [diofant.MatrixSymbol(s, n, n) for s in ["A", "B", "C"]]

In [5]: diofant.Eq(A + B, B + A)
Out[5]: true

In [6]: diofant.Eq(A * (B + C), A * B + A * C)
Out[6]: A⋅(B + C) = A⋅B + A⋅C

In [7]: diofant.expand(A * (B + C))
Out[7]: A⋅(B + C)
```

I believe `diofant.Eq(A * (B + C), A * B + A * C)` should simplify to `true`.

**Comment by skirpichev** [comment]
Yeah, probably there should be ``_eval_expand_mul`` helper for matrix expressions.

**Comment by goretkin** [comment]
I thought I had posted this before, but I'll repost:

I'm trying to understand this issue in the context of https://github.com/diofant/diofant/commit/8221ff3452d5d33919ed42291b521a16f7764b9f . Is there a way to re-use the same code that expands `a*(b+c)` in the scalar case?

**Comment by skirpichev** [comment]
> Is there a way to re-use the same code that expands a*(b+c) in the scalar case?

I doubt it's possible with current code organization.  Basically, we have two Mul classes.  Ordinary Mul and MatMul.  And only first has expand_mul helper.

Please note that that helper works for non-commutative symbols too:
```
In [7]: a = Symbol('a', commutative=False)

In [8]: b = Symbol('b', commutative=False)

In [9]: c = Symbol('c', commutative=False)

In [10]: a*(b + c)
Out[10]: a⋅(b + c)

In [11]: expand_mul(_)
Out[11]: a⋅b + a⋅c
```

**Comment by goretkin** [comment]
I would like to make an attempt at implementing this, but I'm having trouble understanding the code.

I see: https://github.com/diofant/diofant/blob/3db9c8d7bd8dcf1c6f6717e65d2816d238a91840/diofant/core/expr.py#L3043

which I assume then calls: https://github.com/diofant/diofant/blob/5da845ca1611c271141de31eb401da5fb69c3bf2/diofant/core/mul.py#L767

Is the `_eval_expand_mul` the helper you are referring to?

**Comment by skirpichev** [comment]
Correct.  With current code organization, there should be a similar helper for MatMul.

I don't like this approach, there is something wrong.  Perhaps, symbolic arithmetics should be implemented like in [sympycore](https://github.com/pearu/sympycore), see e.g. [this](https://github.com/sympy/sympy/wiki/Algebras-in-SymPyCore).

See also [this thread](https://groups.google.com/forum/#!searchin/sympy/matrix$20expressions$20op_priority|sort:relevance/sympy/ekqyrxezwQ0/W3fg93dSiy0J), maybe this helps you understand current design, where we have Expr and MatrixExpr, Mul and MatMul (derived from MatrixExpr)...

**Comment by skirpichev** [comment]
Probably, porting of https://github.com/sympy/sympy/pull/15373 may solve this issue.  But see https://github.com/sympy/sympy/issues/15665.
